app/controllers/sessions_controller.rb

```diff
@@ -15,4 +22,36 @@ def set_user_locale
   def default_locale?
     cookies[:locale].eql?(I18n.default_locale)
   end
+
+  def check_recaptcha
+    if recaptcha_required_for_user? && !verify_recaptcha
+      self.resource = resource_class.new sign_in_params
+      redirect_to new_session_path(resource_name, recaptcha: true)
+    else
+      cookies.delete(:login_recaptcha)
+    end
+  end
+
+  def set_recaptcha
+    return unless show_recaptcha?
+
+    @display_recaptcha = true
+    cookies.permanent[:login_recaptcha] = true
+  end
+
+  def show_recaptcha?
+    RecaptchaService.recaptcha_configured? &&
+      (recaptcha_required_for_user? ||
+      params[:recaptcha] ||
+      cookies[:login_recaptcha])
```

Comment:
This handles three different scenarios...
1. User enters incorrect password, in which case the the email is still in the request parameters, so we can use `recaptcha_required_for_user?` to look up the number of failed attempts
2. If a user tries to login without filling out the recaptcha (which shouldn't really happen since it's disabled), then we redirect to the `new` action with the `recaptcha` param
3. If a user needs a recaptcha, they'll have the `login_recaptcha` cookie set, so if they leave before successfully signing in, then when they return, we'll know to sure the recaptcha

Admittedly, there is probably a cleaner way to handle these scenarios, so I'm open to any suggestions!